runtime/src/main/java/com/datastax/oss/quarkus/runtime/metrics/MicroProfileMetricsFactory.java

```diff
@@ -0,0 +1,136 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.quarkus.runtime.metrics;
+
+import com.datastax.dse.driver.api.core.metrics.DseNodeMetric;
+import com.datastax.dse.driver.api.core.metrics.DseSessionMetric;
+import com.datastax.oss.driver.api.core.config.DefaultDriverOption;
+import com.datastax.oss.driver.api.core.config.DriverExecutionProfile;
+import com.datastax.oss.driver.api.core.metadata.Node;
+import com.datastax.oss.driver.api.core.metrics.DefaultNodeMetric;
+import com.datastax.oss.driver.api.core.metrics.DefaultSessionMetric;
+import com.datastax.oss.driver.api.core.metrics.Metrics;
+import com.datastax.oss.driver.api.core.metrics.NodeMetric;
+import com.datastax.oss.driver.api.core.metrics.SessionMetric;
+import com.datastax.oss.driver.internal.core.context.InternalDriverContext;
+import com.datastax.oss.driver.internal.core.metrics.MetricsFactory;
+import com.datastax.oss.driver.internal.core.metrics.NodeMetricUpdater;
+import com.datastax.oss.driver.internal.core.metrics.NoopNodeMetricUpdater;
+import com.datastax.oss.driver.internal.core.metrics.NoopSessionMetricUpdater;
+import com.datastax.oss.driver.internal.core.metrics.SessionMetricUpdater;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import io.smallrye.metrics.MetricsRegistryImpl;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import org.apache.commons.lang3.NotImplementedException;
+import org.eclipse.microprofile.metrics.Metric;
+import org.eclipse.microprofile.metrics.MetricRegistry;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class MicroProfileMetricsFactory implements MetricsFactory {
+  private static final Logger LOG = LoggerFactory.getLogger(MicroProfileMetricsFactory.class);
+
+  private final String logPrefix;
+  private final InternalDriverContext context;
+  private final Set<NodeMetric> enabledNodeMetrics;
+  private final MetricRegistry registry;
+  private final SessionMetricUpdater sessionUpdater;
+
+  public MicroProfileMetricsFactory(InternalDriverContext context) {
+    this.logPrefix = context.getSessionName();
+    this.context = context;
+
+    DriverExecutionProfile config = context.getConfig().getDefaultProfile();
+    Set<SessionMetric> enabledSessionMetrics =
+        parseSessionMetricPaths(config.getStringList(DefaultDriverOption.METRICS_SESSION_ENABLED));
+    this.enabledNodeMetrics =
+        parseNodeMetricPaths(config.getStringList(DefaultDriverOption.METRICS_NODE_ENABLED));
+
+    if (enabledSessionMetrics.isEmpty() && enabledNodeMetrics.isEmpty()) {
+      LOG.debug("[{}] All metrics are disabled.", logPrefix);
+      this.registry = null;
+      this.sessionUpdater = NoopSessionMetricUpdater.INSTANCE;
+    } else {
+      this.registry = new MetricsRegistryImpl();
```

Comment:
Yes, I was also thinking about injecting it there, but I thought that this Hdr custom implementation is a blocker. Since you are ok with duplicating `HdrHistogram` I went that path and implemented your suggestions.
Right now the code uses injected `MetricRegistry`: https://github.com/datastax/cassandra-quarkus/pull/3/files#diff-2cda2a7b5c1ea59242c6192fe0b13d95R62
I've also followed the suggestions on the quarkus-extensions docs, and when metrics library is not present, I am injecting `NoopMetricRegistry`:
https://github.com/datastax/cassandra-quarkus/pull/3/files#diff-4d87f287c23fd7e2ebc1d0421519bf60R220

One big advantage of this approach is that node-level metrics now work out-of-the-box.
I've added more tests that cover all metrics (node and session):
https://github.com/datastax/cassandra-quarkus/pull/3/files#diff-2db56229f70156488ab248f0b4822378R56-R93

>SmallRye's DefaultTimer has a public constructor that takes a reservoir. We already know how to write a reservoir backed by HdrHistogram.

Regarding Timer, I think you meant: `TimerImpl` class and `Reservoir`.
I've created a separate JIRA task for this:  
https://datastax-oss.atlassian.net/browse/JAVA-2734